pytorch_lightning/callbacks/timer.py

```diff
@@ -0,0 +1,173 @@
+        self._duration = duration.total_seconds() if duration is not None else None
+        self._interval = interval
+        self._verbose = verbose
+        self._start_time = {stage: None for stage in RunningStage}
+        self._end_time = {stage: None for stage in RunningStage}
+        self._offset = 0
+
+    def start_time(self, stage: str = RunningStage.TRAINING) -> Optional[float]:
+        """Return the start time of a particular stage (in seconds)"""
+        stage = RunningStage(stage)
+        return self._start_time[stage]
+
+    def end_time(self, stage: str = RunningStage.TRAINING) -> Optional[float]:
+        """Return the end time of a particular stage (in seconds)"""
+        stage = RunningStage(stage)
+        return self._end_time[stage]
+
+    def time_elapsed(self, stage: str = RunningStage.TRAINING) -> float:
+        """Return the time elapsed for a particular stage (in seconds)"""
+        start = self.start_time(stage)
+        end = self.end_time(stage)
+        offset = self._offset if stage == RunningStage.TRAINING else 0
+        if start is None:
+            return offset
+        if end is None:
+            return time.monotonic() - start + offset
+        return end - start + offset
+
+    def time_remaining(self, stage: str = RunningStage.TRAINING) -> Optional[float]:
+        """Return the time remaining for a particular stage (in seconds)"""
+        if self._duration is not None:
+            return self._duration - self.time_elapsed(stage)
+
+    def on_train_start(self, *args, **kwargs) -> None:
+        self._start_time[RunningStage.TRAINING] = time.monotonic()
+
+    def on_train_end(self, *args, **kwargs) -> None:
+        self._end_time[RunningStage.TRAINING] = time.monotonic()
+
+    def on_validation_start(self, *args, **kwargs) -> None:
+        self._start_time[RunningStage.VALIDATING] = time.monotonic()
+
+    def on_validation_end(self, *args, **kwargs) -> None:
+        self._end_time[RunningStage.VALIDATING] = time.monotonic()
+
+    def on_test_start(self, *args, **kwargs) -> None:
+        self._start_time[RunningStage.TESTING] = time.monotonic()
+
+    def on_test_end(self, *args, **kwargs) -> None:
+        self._end_time[RunningStage.TESTING] = time.monotonic()
+
+    def on_train_batch_end(self, trainer: 'pl.Trainer', *args, **kwargs) -> None:
+        if self._interval != Interval.step or self._duration is None:
+            return
+        self._check_time_remaining(trainer)
+
+    def on_train_epoch_end(self, trainer: 'pl.Trainer', *args, **kwargs) -> None:
+        if self._interval != Interval.epoch or self._duration is None:
+            return
+        self._check_time_remaining(trainer)
+
+    def on_save_checkpoint(
+        self,
+        trainer: 'pl.Trainer',
+        pl_module: 'pl.LightningModule',
+        checkpoint: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        return {"time_elapsed": {stage.value: self.time_elapsed(stage) for stage in list(RunningStage)}}
```

Comment:
@carmocca I'm changing this back to value, because otherwise we save the keys as 
`<RunningStage.TRAINING: 'train'>` etc. instead of just "train"